# Code-Profiling
This notebook shows 3 different ways of using [line_profiler](https://github.com/rkern/line_profiler#line-profiler) to profile the runtime of Python code line by line.

## Runtime profiling with line_profiler

### Using line_profiler within session

In [1]:
from line_profiler import LineProfiler

In [74]:
def function_1(n):
    for i in range(n):
        x = [5 * j for j in range(10000)]
        y = function_2(n)
    
def function_2(n):
    s1 = [i for i in range(n)]
    s2 = sum(s1)

In [73]:
lp = LineProfiler()
lp_wrapper = lp(function_1)
lp.add_function(function_2)   # add additional function to profile
lp_wrapper(n = 1000)
lp.print_stats()

Timer unit: 1e-06 s

Total time: 1.48845 s
File: <ipython-input-70-5a497711d596>
Function: function_1 at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           def function_1(n):
     2      1001        385.0      0.4      0.0      for i in range(n):
     3      1000    1381318.0   1381.3     92.8          x = [5 * j for j in range(10000)]
     4      1000     106748.0    106.7      7.2          y = function_2(n)

Total time: 0.100231 s
File: <ipython-input-70-5a497711d596>
Function: function_2 at line 6

Line #      Hits         Time  Per Hit   % Time  Line Contents
     6                                           def function_2(n):
     7      1000      95154.0     95.2     94.9      s1 = [i for i in range(n)]
     8      1000       5077.0      5.1      5.1      s2 = sum(s1)



### %lprun magic function
First we must load the line_profiler Ipython extension;

In [19]:
%load_ext line_profiler

Note that the output of %lprun is displayed in a sub window in the browser so has been copied into a markdown cell here.

In [46]:
%lprun -f do_stuff -f do_other_stuff do_stuff(numbers)

Timer unit: 1e-06 s

Total time: 0.004954 s
File: <ipython-input-10-5958c14c4b77>
Function: do_other_stuff at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents 
==============================================================
     1                                           def do_other_stuff(numbers):
     2      1001       4954.0      4.9    100.0      s = sum(numbers)

Total time: 0.59044 s
File: <ipython-input-10-5958c14c4b77>
Function: do_stuff at line 4

Line #      Hits         Time  Per Hit   % Time  Line Contents 
==============================================================
     4                                           def do_stuff(numbers):
     5      1001        582.0      0.6      0.1      for i in range(1000):
     6      1000       6054.0      6.1      1.0          do_other_stuff(numbers)
     7      1000     177508.0    177.5     30.1          l = [numbers[i]/43 for i in range(len(numbers))]
     8      1000     406290.0    406.3     68.8          m = ['hello'+str(numbers[i]) for i in range(len(numbers))]
     9         1          6.0      6.0      0.0      do_other_stuff(numbers)

### From command line
It is also possible to use line_profiler from the command line. The functions above have been saved in a file 'functions_to_profile.py';

In [76]:
with open('functions_to_profile.py', 'r') as viewFileOpen:
        data = viewFileOpen.read()
print(data)

@profile
def function_1(n):
    for i in range(n):
        x = [5 * j for j in range(10000)]
        y = function_2(n)
    
@profile
def function_2(n):
    s1 = [i for i in range(n)]
    s2 = sum(s1)


function_1(1000)



Now we use the %system magic command to run a command line command in jupyter and capture the results;

In [77]:
%system kernprof -l -v functions_to_profile.py

['Wrote profile results to functions_to_profile.py.lprof',
 'Timer unit: 1e-06 s',
 '',
 'Total time: 1.54464 s',
 'File: functions_to_profile.py',
 'Function: function_1 at line 1',
 '',
 'Line #      Hits         Time  Per Hit   % Time  Line Contents',
 '==============================================================',
 '     1                                           @profile',
 '     2                                           def function_1(n):',
 '     3      1001        412.0      0.4      0.0      for i in range(n):',
 '     4      1000    1428096.0   1428.1     92.5          x = [5 * j for j in range(10000)]',
 '     5      1000     116136.0    116.1      7.5          y = function_2(n)',
 '',
 'Total time: 0.107712 s',
 'File: functions_to_profile.py',
 'Function: function_2 at line 7',
 '',
 'Line #      Hits         Time  Per Hit   % Time  Line Contents',
 '==============================================================',
 '     7                                           @pr